### PR TITLE
Simplify API

### PR DESCRIPTION
**Pull request by NathanReb**
This PR simplifies the fiat API. This is a follow up to some feedback from @hannesm.

I extracted `generate_key` and `key_echange` functions to a `Dhe` submodule.

We could simplify it further by renaming some of the types ie `point` -> `public_key` and `scalar` -> `private_key` but that's out of the scope of this PR.

### CODE REVIEW

File: p256/fiat_p256.mli [hunk 1]
```diff
@@ -48,14 +25,19 @@ val scalar_of_cs : Cstruct.t -> (scalar, scalar_error) result
...
-val dh : scalar:scalar -> point:point -> Cstruct.t
-(** Perform Diffie-Hellman key exchange. This returns the x component of the
-    scalar multiplication of [point] and [scalar]. *)
-
-val public : scalar -> point
-(** Compute the public key corresponding to a given private key. Internally,
-    this multiplies the generator by the scalar.
-    Given the invariant on [scalar], the result can't be the point at infinity. *)
+module Dhe : sig
+  val generate_key : rng:(int -> Cstruct.t) -> scalar * Cstruct.t
```
[the remark this answers is not in the document]
Comment (NathanReb) [reply p256/fiat_p256.mli]: I personally would like to label it but I seem to recall someone mentioning this was some standard API for RNG functions in mirage somewhere. I'll investigate and add a label if it's not!

[the remark this answers is not in the document]
Comment (hannesm) [reply p256/fiat_p256.mli]: https://docs.mirage.io/mirage-random/Mirage_random/module-type-S/index.html#val-generate

[the remark this answers is not in the document]
Comment (dinosaure) [reply p256/fiat_p256.mli]:
```ocaml
val generate_key : ?g:'g -> rng:(?g:'g -> int -> Cstruct.t) -> scalar * Cstruct.t
```

should fit under common interface `Mirage_random.S` (specialized on `Cstruct.t`).

**Comment by hannesm** [comment]
"We could simplify it further by renaming some of the types ie point -> public_key and scalar -> private_key but that's out of the scope of this PR." so what is then in scope for this PR? I still fail to find an answer how fiat and hacl APIs correlate (in terms of type names and function signatures), and something about the size of the signature.

if you compare https://github.com/mirage/hacl/pull/14 with this, there's still quite some differences, esp. in names of types and labels. is this intentional? why is there a `Dhe` submodule in this PR, where is it useful? why is the error variant called `point_error` (and `pp_point_error`) -- it's now only used in `key_exchange` afaict?

**Comment by NathanReb** [comment]
> so what is then in scope for this PR?

I wanted to expose the signature you asked for first as that's the most important part, the rest is just a matter of renaming types. I wanted to keep PRs small but I get that you'd prefer to just look at the final API.

> why is there a Dhe submodule in this PR, where is it useful?

I think putting those functions in a `Dhe` module is more future proof and allows us to eventually expose different APIs for signature or non-ephemeral diffie-hellman which need to reuse public keys and thus might require a different API that uses intermediate types instead of raw `Cstruct.t`.
I find it clearer to use submodules than putting affixes in functions and type names such as `generate_dhe_key` and `dhe_key_exchange` but that might be subjective.

The whole idea is to make the API extensible with other EC based crypto operation and isolate them from one another.

File: p256/fiat_p256.mli (hunk 1 above)
[the remark this answers is not in the document]
Comment (NathanReb) [reply p256/fiat_p256.mli]: I sticked to the current `rng: (int -> Cstruct.t)` as it's more generic and still compatible with `Mirage_random.S` without too much hassle.

File: p256/fiat_p256.mli [hunk 2]
```diff
@@ -1,61 +1,31 @@
...
-    It should be 32 bytes long, in big endian format. Returns an error when the
-    number is zero, or if it is larger than or equal to the group order. *)
+    [rng] is the function used to repeteadly generate a private key until a valid candidate
+    is obtained. [rng]'s int parameter is the size of the [Cstruct.t] to generate.
+    If [rng] returns an invalid length buffer, [Failure _] is raised.
 
-val scalar_of_hex : Hex.t -> (scalar, scalar_error) result
-(** Like [scalar_of_cs] but read from hex data. *)
+    The generated private key is checked to be greater than zero and lower than the group
+    order meaning the public key cannot be the point at inifinity. *)
```
Comment (hannesm) [inline p256/fiat_p256.mli]: please include a comment about the second projection of the returned value -- is it truncated if there are leading 0 bytes? (i suspect no, which is fine (needed in TLS 1.3, whereas truncation is require in TLS 1.2 and earlier)).

File: p256/fiat_p256.mli [hunk 3]
```diff
@@ -1,61 +1,31 @@
...
 
-val dh : scalar:scalar -> point:point -> Cstruct.t
-(** Perform Diffie-Hellman key exchange. This returns the x component of the
-    scalar multiplication of [point] and [scalar]. *)
+val key_exchange : secret -> Cstruct.t -> (Cstruct.t, error) result
+(** [key_exchange ~private_key received_public_key] performs Diffie-Hellman key exchange
```
Comment (hannesm) [inline p256/fiat_p256.mli]:
```suggestion
(** [key_exchange secret received] performs Diffie-Hellman key exchange
```

File: p256/fiat_p256.mli [hunk 4]
```diff
@@ -1,61 +1,31 @@
...
+    using your private key and the other party's public key. Returns the shared secret
```
Comment (hannesm) [inline p256/fiat_p256.mli]:
```suggestion
    using your secret and the received data from the other party. Returns the shared secret
```

File: p256/fiat_p256.mli [hunk 5]
```diff
@@ -1,61 +1,31 @@
...
+    or an error if the received public is invalid or is the point at infinity.
 
-val public : scalar -> point
-(** Compute the public key corresponding to a given private key. Internally,
-    this multiplies the generator by the scalar.
-    Given the invariant on [scalar], the result can't be the point at infinity. *)
+    @see <http://www.secg.org/sec1-v2.pdf> for public key encoding format. *)
```
Comment (hannesm) [inline p256/fiat_p256.mli]: here as well, please mention truncation (i suspect the shared secret is not truncated).

what is an `invalid` public (i suspect bad encoding, not on curve)?

**Comment by hannesm** [comment]
looks good to me, i added minor comments regarding the docstrings in the interface. thanks!

File: p256/fiat_p256.mli (hunk 2 above)
Comment (NathanReb) [reply p256/fiat_p256.mli]: I'm not sure which truncation you're talking about? It shouldn't matter for the `secret` as it's an abstract type so I'm guessing you mean for the returned encoded public key in the Cstruct but I'm unsure what there is to truncate there.

Comment (hannesm) [reply p256/fiat_p256.mli]: sorry, its too warm -- i was thinking about https://tools.ietf.org/html/rfc5246#section-4.4 -- but it looks like mainly leading 0 bytes of the computed shared secret Z are stripped (see https://tools.ietf.org/html/rfc5246#section-8.1.2) (that is tls 1.2, 1.3 is different in that behaviour).

Comment (NathanReb) [reply p256/fiat_p256.mli]: No worries ! I updated the `key_exchange` documentation according to your other similar comment !

**Pull request merged**